src/nlp.jl

```diff
@@ -1652,53 +1652,53 @@ function MOI.constraint_expr(d::NLPEvaluator, i::Integer)
     end
 end
 
+
+"""
+    _VarValueMap{T,F}
+
+A lazy cache used for computing the primal variable solution in `forward_eval`.
+
+This avoids the need to rewrite the nonlinear expressions from MOIVARIABLE to
+VARIABLE, as well as eagerly computing the `var_value` for every variable. We
+use a `cache` so we don't have to recompute variables we have already seen.
+"""
+struct _VarValueMap{T,F} <: AbstractVector{T}
```

Comment:
It's better to remove the restriction on `x_values` being an `AbstractVector` rather than faking it here.